cmd/agent/gui/checks.go

```diff
@@ -243,42 +250,98 @@ func listFiles(w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(res))
 }
 
-// Helper function which returns the appropriate paths for finding checks/check configs
-func getAllPaths(fileType string) []string {
-	if fileType == "py" {
-		return []string{
-			filepath.Join(common.GetDistPath(), "checks.d"), // Custom checks
-			config.Datadog.GetString("additional_checksd"),  // Custom checks
-			common.PyChecksPath,                             // Integrations-core checks
-		}
-	} else if fileType == "yaml" {
-		return []string{
-			config.Datadog.GetString("confd_path"),        // Custom checks
-			filepath.Join(common.GetDistPath(), "conf.d"), // Default check configs
+// Sends a list containing the names of all the config files
+func listConfigs(w http.ResponseWriter, r *http.Request) {
+	filenames := []string{}
+	for _, path := range configPaths {
+		files, e := readConfDir(path)
+
+		if e == nil {
+			// If a default config is found but a non-default version exists, ignore default
+			sort.Strings(files)
+			lookup := make(map[string]bool)
+			for _, name := range files {
+
+				var checkName string
+				if i := strings.Index(name, ":"); i != -1 {
+					// If the check was in a subdir, '{checkname}: ' prepends the filename
+					checkName = name[:i]
+				} else {
+					checkName = name[:strings.Index(name, ".")]
+				}
+
+				if ext := filepath.Ext(name); ext == ".default" {
+					if _, exists := lookup[checkName]; !exists {
+						filenames = append(filenames, name)
+						lookup[checkName] = true
+					}
+					continue
+				}
+
+				filenames = append(filenames, name)
+				lookup[checkName] = true
+			}
 		}
 	}
-	return []string{}
+
+	if len(filenames) == 0 {
+		w.Write([]byte("No configuration (.yaml) files found."))
+		return
+	}
+
+	res, _ := json.Marshal(filenames)
+	w.Header().Set("Content-Type", "application/json")
+	w.Write([]byte(res))
 }
 
-// Helper function which returns all the filenames in a directory
-func readDirectory(path string) ([]string, error) {
+// Helper function which returns all the filenames in a check config directory
+func readConfDir(path string) ([]string, error) {
 	var filenames []string
-	dir, e := os.Open(path)
-	if e != nil {
-		return filenames, e
+	entries, err := ioutil.ReadDir(path)
+	if err != nil {
+		return filenames, err
 	}
-	defer dir.Close()
 
-	files, e := dir.Readdir(-1)
-	if e != nil {
-		return filenames, e
-	}
+	for _, entry := range entries {
+		// Some check configs are in nested subdirectories
+		if entry.IsDir() {
+			if filepath.Ext(entry.Name()) != ".d" {
+				continue
+			}
+
+			subEntries, err := ioutil.ReadDir(path + "/" + entry.Name())
+			if err == nil {
+				checkName := strings.TrimSuffix(entry.Name(), ".d")
+
+				for _, subEntry := range subEntries {
+					if hasRightEnding(subEntry.Name()) && subEntry.Mode().IsRegular() {
+						filenames = append(filenames, checkName+": "+subEntry.Name())
```

Comment:
We should use a struct instead of concatenating the check name and the file. It forces use the split the string in every other place in the code.

Reply:
The logic here was that this string is the 'display name' that will be shown on the front end, so it makes sense for it to be a properly formatted string rather than a JSON object when the front end receives it... the helper function I added (as suggested above) should mitigate this issue, but definitely let me know if you think that isn't sufficient!